sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/DeduplicateRelations.scala

```diff
@@ -17,16 +17,30 @@
 
 package org.apache.spark.sql.catalyst.analysis
 
-import scala.collection.mutable.ArrayBuffer
+import scala.collection.mutable
 
 import org.apache.spark.sql.catalyst.expressions.{Alias, AttributeMap, AttributeSet, NamedExpression, SubqueryExpression}
 import org.apache.spark.sql.catalyst.plans.logical._
 import org.apache.spark.sql.catalyst.rules.Rule
 import org.apache.spark.sql.catalyst.trees.AlwaysProcess
 
+/**
+ * A help class that used to detect duplicate relations fast in `DeduplicateRelations`
+ */
+case class ReferenceEqualPlanWrapper(plan: LogicalPlan) {
+  private val _hashCode = System.identityHashCode(plan)
```

Comment:
Why using System.identityHashCode instead of plan.hash()?

Reply:
We only want to know if the two plans are the same instance. By default, `System.identityHashCode` returns the memory address of an object, which is faster than plan.hash() (which involves more details about the plan itself.)

Reply:
Ah, I see. The java library does't have `IdentityHasSet`, even though it has `IdentityHasMap`...

Reply:
Yea..I actually tried `IdentityHasMap` too but it doesn't help a lot...